docs/root/configuration/http_filters/original_src_filter.rst

```diff
@@ -0,0 +1,65 @@
+.. _config_http_filters_original_src:
+
+Original Source
+===============
+
+* :ref:`HTTP filter v2 API reference <envoy_api_msg_config.filter.http.original_src.v2alpha1.OriginalSrc>`
+* This filter should be configured with the name *envoy.filters.http.original_src*.
+
+The original source http filter replicates the downstream remote address of the connection on
+the upstream side of Envoy. For example, if a downstream connection connects to Envoy with IP
+address ``10.1.2.3``, then Envoy will connect to the upstream with source IP ``10.1.2.3``. The
+downstream remote address is determined based on the logic for the "trusted client address"
+outlined in :ref:`XFF <config_http_conn_man_headers_x-forwarded-for>`.
+
+
+IP Version Support
+------------------
+The filter supports both IPv4 and IPv6 as addresses. Note that the upstream connection must support
+the version used.
+
+Extra Setup
+-----------
+
+The downstream remote address used will likely be globally routable. By default, packets returning
+from the upstream host to that address will not route through Envoy. The network must be configured
+to forcefully route any traffic whose IP was replicated by Envoy back through the Envoy host.
+
+If Envoy and the upstream are on the same host -- e.g. in an sidecar deployment --, then iptables
+and routing rules can be used to ensure correct behaviour. The filter has an unsigned integer
+configuration,
+:ref:`mark <envoy_api_field_config.filter.http.original_src.v2alpha1.OriginalSrc.mark>`. Setting
+this to *X* causes Envoy to *mark* all upstream packets originating from this http with value
+*X*. Note that if
+:ref:`mark <envoy_api_field_config.filter.http.original_src.v2alpha1.OriginalSrc.mark>` is set
+to 0, Envoy will not mark upstream packets.
+
+We can use the following set of commands to ensure that all ipv4 and ipv6 traffic marked with *X*
+(assumed to be 123 in the example) routes correctly. Note that this example assumes that *eth0* is
+the default outbound interface.
+
+.. code-block:: text
+
+  iptables  -t mangle -I PREROUTING -m mark     --mark 123 -j CONNMARK --save-mark
+  iptables  -t mangle -I OUTPUT     -m connmark --mark 123 -j CONNMARK --restore-mark
+  ip6tables -t mangle -I PREROUTING -m mark     --mark 123 -j CONNMARK --save-mark
+  ip6tables -t mangle -I OUTPUT     -m connmark --mark 123 -j CONNMARK --restore-mark
+  ip rule add fwmark 123 lookup 100
+  ip route add local 0.0.0.0/0 dev lo table 100
+  ip -6 rule add fwmark 123 lookup 100
+  ip -6 route add local ::/0 dev lo table 100
+  echo 1 > /proc/sys/net/ipv4/conf/eth0/route_localnet
+
+
+Example HTTP configuration
+------------------------------
+
+The following example configures Envoy to use the original source for all connections made on port
+8888. All upstream packets are marked with 123.
+
+.. code-block:: yaml
+
+  http_filters:
```

Comment:
Should this be used in conjunction with the router filter? Or does it replace it? Might be good to be explicit about how this is supposed to work in the docs

Reply:
Good point. It's in conjunction, and should be *before* (I think... I haven't tried with it after, but my understanding is that it wouldn't make sense otherwise).

Reply:
Yeah having it after makes no sense, the router terminates the filter chain

Reply:
I'll document that it's meant to be used to inform the router, rather than instead of, and that it should therefore appear earlier in the chain. I'll also update the example to make it clear.